File: BFS/BFS/BFS.cpp
#include<iostream>
#include <list>
using namespace std;
//实现图
class Graphic {
public:
	Graphic(int vertexNum);
	void BFS(int ver);
	void addEdge(int v,int w);
	~Graphic();
private:
	int V;
	list<int>* adj;
};
Graphic::Graphic(int V):V(V) {
	adj = new list<int>[V];
}
void Graphic::addEdge(int v, int w) {
	adj[v].push_back(w);
}
void Graphic::BFS(int ver) {
	list<int> queue;
	bool* visited = new bool[V];
	for (int i = 0; i < V; i++) {
		visited[i] = false;
	}
	visited[ver] = true;
	queue.push_back(ver);
	while (!queue.empty()) {
		int s = queue.front();
		cout << s << " ";
		queue.pop_front();

		list<int>::iterator v;
		for (v = adj[s].begin(); v != adj[s].end(); v++) {
			if (!visited[*v]) {
				visited[*v] = true;
				queue.push_back(*v);
			}
		}


	}
}
Graphic::~Graphic() {
	delete []adj;
}
int main() {
	int nums;
	cout << "请输入顶点数:";
	cin >> nums;
	Graphic graphic(nums);
	int numVer;
	cout << "请输入边：" << endl;
	cin >> numVer;
	for (int i=0;i<numVer;i++){
		int ver1, ver2;
		cin >> ver1;
		cin >> ver2;
		graphic.addEdge(ver1, ver2);
	}
	graphic.BFS(2);
	system("pause");
	return 0;
}

File: py_BFS/BFS.py
from collections import defaultdict
class Graphic:
    def __init__(self):
        self.adj=defaultdict(list)
    def addEdge(self,v,w):
        self.adj[v].append(w)
    def BFS(self,ver):
        visited=[False]*(len(self.adj))
        queue=[]
        visited[ver]=True
        queue.append(ver)
        
        while queue:
            s=queue.pop(0)
            print(s,end=' ')
            for i in self.adj[s]:
                if visited[i] is False:
                    visited[i]=True
                    queue.append(i)
numEdge=int(input("请输入边数："))
g=Graphic()
for i in range(numEdge):
    edge=input("请输入一条边：")
    vert=[int (x) for x in edge.split()]
    g.addEdge(vert[0],vert[1])
g.BFS(2)

